tools/gyp/pylib/gyp/common.py

```diff
@@ -341,148 +348,146 @@ def WriteOnDiff(filename):
-      except Exception:
-        # Don't leave turds behind.
-        os.unlink(self.tmp_path)
-        raise
-
-    def __getattr__(self, attrname):
-      # Delegate everything else to self.tmp_file
-      return getattr(self.tmp_file, attrname)
-
-    def close(self):
-      try:
-        # Close tmp file.
-        self.tmp_file.close()
-        # Determine if different.
-        same = False
-        try:
-          same = filecmp.cmp(self.tmp_path, filename, False)
-        except OSError as e:
-          if e.errno != errno.ENOENT:
-            raise
-
-        if same:
-          # The new file is identical to the old one, just get rid of the new
-          # one.
-          os.unlink(self.tmp_path)
-        else:
-          # The new file is different from the old one, or there is no old one.
-          # Rename the new file to the permanent name.
-          #
-          # tempfile.mkstemp uses an overly restrictive mode, resulting in a
-          # file that can only be read by the owner, regardless of the umask.
-          # There's no reason to not respect the umask here, which means that
-          # an extra hoop is required to fetch it and reset the new file's mode.
-          #
-          # No way to get the umask without setting a new one?  Set a safe one
-          # and then set it back to the old value.
-          umask = os.umask(0o77)
-          os.umask(umask)
-          os.chmod(self.tmp_path, 0o666 & ~umask)
-          if sys.platform == 'win32' and os.path.exists(filename):
-            # NOTE: on windows (but not cygwin) rename will not replace an
-            # existing file, so it must be preceded with a remove. Sadly there
-            # is no way to make the switch atomic.
-            os.remove(filename)
-          os.rename(self.tmp_path, filename)
-      except Exception:
-        # Don't leave turds behind.
-        os.unlink(self.tmp_path)
-        raise
-
-    def write(self, s):
-      self.tmp_file.write(s.encode('utf-8'))
-
-  return Writer()
+    class Writer(object):
+        """Wrapper around file which only covers the target if it differs."""
+
+        def __init__(self):
+            # On Cygwin remove the "dir" argument because `C:` prefixed paths are treated as relative,
+            # consequently ending up with current dir "/cygdrive/c/..." being prefixed to those, which was
+            # obviously a non-existent path, for example: "/cygdrive/c/<some folder>/C:\<my win style abs path>".
+            # See https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp for more details
+            base_temp_dir = "" if IsCygwin() else os.path.dirname(filename)
+            # Pick temporary file.
+            tmp_fd, self.tmp_path = tempfile.mkstemp(
+                suffix=".tmp",
+                prefix=os.path.split(filename)[1] + ".gyp.",
+                dir=base_temp_dir,
+            )
+            try:
+                self.tmp_file = os.fdopen(tmp_fd, "w")
```

Comment:
... and here we are opening in `w` mode.

Reply:
Is it wrong?
It came from [this upstream commit](https://chromium.googlesource.com/external/gyp/+/ab4aca868d68b9de9b20a9991bbbb5e78ab48a30%5E!) that was ported in https://github.com/nodejs/gyp-next/pull/11

Reply:
I am unsure but it would explain the `TypeError: write() argument must be str, not byte` above.

Reply:
@cclauss @targos @gengjiawen could someone point me to the smallest configuration that causes gyp to fail? We could simply add a failing test and patch the relevant line highlighted by @cclauss.

Reply:
Looks like `gyp-next` has not been add windows to CI, we can make this as a start ?

Reply:
@gengjiawen indeed, it has been blocked on the test failures in https://github.com/nodejs/gyp-next/pull/8. I'll try to get it resolved ASAP.